using_restricted_array.rb

```diff
@@ -7,50 +7,114 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  length = 0
+  i = 0
+
+  while array[i] != nil
+    length = length + 1
+    i = i + 1
+  end
+  return length
 end
 
-# Prints each integer values in the array
+
+#Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while array[i] != nil
+    print "#{array[i]}"
+    if array[i+1] != nil
+      print ', '
+    end
+    i = i + 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    end
+    i = i + 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  largest_value = array[0]
+  while i < length
+    if array[i] > largest_value
+      largest_value = array[i]
+    end
+    i = i + 1
+  end
+  return largest_value
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  smallest_value = array[0]
+  while i < length
+    if array[i] < smallest_value
+      smallest_value = array[i]
+    end
+    i = i + 1
+  end
+  return smallest_value
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  length = length - 1
+  while i < length
+    last = array[length]
+    first = array[i]
+
+    array[i] = last
+    array[length] = first
+
+    i = i + 1
+    length = length - 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def binary_search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  low = 0
+  high = length
+
+  while low < high
+    mid = (low + high)/2
+    if value_to_find == array[mid]
+      return true
+    elsif array[mid] > value_to_find
+      high = mid - 1
+    elsif array[mid] < value_to_find
+      low = mid + 1
+    end
+  end
```

Comment:
To account for the scenario when there is only one element in the array and it is the value_to_find, add an additional conditional statement before or after the loop to compare array[low] with value_to_find.